### PR TITLE
Fix a bug in the encrypt sample, that was causing the decrypt to fail

### LINKED ISSUE #23814: Encryption example does not work
https://docs.microsoft.com/en-us/dotnet/standard/security/encrypting-data
This encryption example does not work when decrypted with the same key on the page
https://docs.microsoft.com/en-us/dotnet/standard/security/decrypting-data
When decrypting, an empty string is obtained.

```
using System;
using System.IO;
using System.Security.Cryptography;

public class EncryptExample
{
    public static void Main(string[] args)
    {
        //Encryption key used to encrypt the stream.
        //The same value must be used to encrypt and decrypt the stream.
        byte[] key = { 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16 };

        try
        {
            //Create a file stream
            using FileStream myStream = new FileStream("TestData.txt", FileMode.OpenOrCreate);

            //Create a new instance of the default Aes implementation class  
            // and configure encryption key.  
            using Aes aes = Aes.Create();
            aes.Key = key;

            //Stores IV at the beginning of the file.
            //This information will be used for decryption.
            byte[] iv = aes.IV;
            myStream.Write(iv, 0, iv.Length);

            //Create a CryptoStream, pass it the FileStream, and encrypt
            //it with the Aes class.  
            using CryptoStream cryptStream = new CryptoStream(
                myStream, 
                aes.CreateEncryptor(),
                CryptoStreamMode.Write);

            //Create a StreamWriter for easy writing to the
            //file stream.  
            using StreamWriter sWriter = new StreamWriter(cryptStream);

            //Write to the stream.  
            sWriter.WriteLine("Hello World!");

            cryptStream.FlushFinalBlock();

            //Inform the user that the message was written  
            //to the stream.  
            Console.WriteLine("The file was encrypted.");
        }
        catch
        {
            //Inform the user that an exception was raised.  
            Console.WriteLine("The encryption failed.");
            throw;
        }
    }
}
```

instead
```
//Create a StreamWriter for easy writing to the
//file stream.  
using StreamWriter sWriter = new StreamWriter(cryptStream);

//Write to the stream.  
sWriter.WriteLine("Hello World!");
```
I used
```
var inputByteArray = Encoding.UTF8.GetBytes("Hello World!");
cryptStream.Write(inputByteArray, 0, inputByteArray.Length);
```
then the decrypted data is identical.

This may be wrong, but in any case, please check.

Thank you.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: ee15f916-5406-85fa-b7fb-a9e69fcd3efb
* Version Independent ID: b6aae569-8a56-0a44-7b8e-e54491418baf
* Content: [Encrypting data](https://docs.microsoft.com/en-us/dotnet/standard/security/encrypting-data)
* Content Source: [docs/standard/security/encrypting-data.md](https://github.com/dotnet/docs/blob/main/docs/standard/security/encrypting-data.md)
* Product: **dotnet**
* Technology: **dotnet-security**
* GitHub Login: @IEvangelist
* Microsoft Alias: **dapine**

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

- Upgrade encrypt / decrypt snippets to .NET 5
- Use top-level statements
- Use target-type new expressions
- Use explicitly scoped `using` to avoid encryption errors

Fixes #23814

### Internal preview

✔️ [Encrypting data](https://review.docs.microsoft.com/en-us/dotnet/standard/security/encrypting-data?branch=pr-en-us-23816)
✔️ [Decrypting data](https://review.docs.microsoft.com/en-us/dotnet/standard/security/decrypting-data?branch=pr-en-us-23816)

### CODE REVIEW

File: docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs [hunk 1]
```diff
@@ -2,43 +2,41 @@
...
-            Console.WriteLine("The decryption failed.");
-            throw;
-        }
-    }
+        0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
+        0x09, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16
+    };
+
+    //Create a CryptoStream, pass it the file stream, and decrypt
+    //it with the Aes class using the key and IV.
```
Comment (Youssef1313) [inline docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs:25]: nit:

```suggestion
    // Create a CryptoStream, pass it the file stream, and decrypt
    // it with the Aes class using the key and IV.
```

File: docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs [hunk 2]
```diff
@@ -2,43 +2,41 @@
...
+    // Read the stream.
+    using StreamReader decryptReader = new(cryptoStream);
+
+    var decryptedMessage = await decryptReader.ReadToEndAsync();
+
+    Console.WriteLine($"The decrypted original message: {decryptedMessage}");
+}
+catch
+{
+    Console.WriteLine("The decryption failed.");
```
Comment (Youssef1313) [inline docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs:40]: Consider adding the exception message.

File: docs/standard/security/snippets/decrypting-data/vb/aes-decrypt.vb [hunk 3]
```diff
@@ -4,30 +4,30 @@ Imports System.Security.Cryptography
...
-        'to encrypt the stream.
+        ' Decryption key must be the same value that was used
+        ' to encrypt the stream.
         Dim key As Byte() = {&H1, &H2, &H3, &H4, &H5, &H6, &H7, &H8, &H9, &H10, &H11, &H12, &H13, &H14, &H15, &H16}
 
         Try
-            'Create a file stream.
-            Using myStream As FileStream = New FileStream("TestData.txt", FileMode.Open)
+            ' Create a file stream.
+            Using fileStream As FileStream = New FileStream("TestData.txt", FileMode.Open)
```
Comment (Youssef1313) [inline docs/standard/security/snippets/decrypting-data/vb/aes-decrypt.vb:13]:
```suggestion
            Using fileStream As New FileStream("TestData.txt", FileMode.Open)
```

File: docs/standard/security/snippets/encrypting-data/vb/aes-encrypt.vb [hunk 4]
```diff
@@ -4,33 +4,36 @@ Imports System.Security.Cryptography
...
     Sub Main()
-        'Encryption key used to encrypt the stream.
-        'The same value must be used to encrypt and decrypt the stream.
-        Dim key As Byte() = {&H1, &H2, &H3, &H4, &H5, &H6, &H7, &H8, &H9, &H10, &H11, &H12, &H13, &H14, &H15, &H16}
-
         Try
-            'Create a file stream
-            Using myStream As FileStream = New FileStream("TestData.txt", FileMode.OpenOrCreate)
+            ' Create a file stream
+            Using fileStream As FileStream = New FileStream("TestData.txt", FileMode.OpenOrCreate)
```
Comment (Youssef1313) [inline docs/standard/security/snippets/encrypting-data/vb/aes-encrypt.vb:9]:
```suggestion
            Using fileStream As New FileStream("TestData.txt", FileMode.OpenOrCreate)
```

File: docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs (hunk 2 above)
Comment (IEvangelist) [inline docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs:40]:
```suggestion
catch (Execption ex)
{
    Console.WriteLine($"The decryption failed. {ex}");
```

File: docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs [hunk 5]
```diff
@@ -2,43 +2,41 @@
 using System.IO;
```
Comment (IEvangelist) [inline docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs:2]:
```suggestion
using System;
using System.IO;
```

File: docs/standard/security/snippets/encrypting-data/csharp/aes-encrypt.cs [hunk 6]
```diff
@@ -2,54 +2,48 @@
...
         }
     }
+
+    Console.WriteLine("The file was encrypted.");
+}
+catch
+{
+    //Inform the user that an exception was raised.
+    Console.WriteLine("The encryption failed.");
+    throw;
```
Comment (IEvangelist) [inline docs/standard/security/snippets/encrypting-data/csharp/aes-encrypt.cs:48]:
```suggestion
catch (Exception ex)
{
    Console.WriteLine($"The encryption failed. {ex}");
```

File: docs/standard/security/snippets/encrypting-data/csharp/aes-encrypt.cs [hunk 7]
```diff
@@ -2,54 +2,48 @@
 using System.IO;
```
Comment (IEvangelist) [inline docs/standard/security/snippets/encrypting-data/csharp/aes-encrypt.cs:2]:
```suggestion
using System;
using System.IO;
```

File: docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs [hunk 8]
```diff
@@ -1,44 +1,43 @@
 ﻿using System;
+using System;
```
Comment (IEvangelist) [inline docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs:2]:
```suggestion
```

File: docs/standard/security/snippets/encrypting-data/csharp/aes-encrypt.cs [hunk 9]
```diff
@@ -1,55 +1,48 @@
 ﻿using System;
+using System;
```
Comment (IEvangelist) [inline docs/standard/security/snippets/encrypting-data/csharp/aes-encrypt.cs:2]:
```suggestion
```

File: docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs [hunk 10]
```diff
@@ -2,43 +2,41 @@
...
+       CryptoStreamMode.Read);
+
+    // Read the stream.
+    using StreamReader decryptReader = new(cryptoStream);
+
+    var decryptedMessage = await decryptReader.ReadToEndAsync();
+
+    Console.WriteLine($"The decrypted original message: {decryptedMessage}");
+}
+catch (Execption ex)
```
Comment (IEvangelist) [inline docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs:38]:
```suggestion
catch (Exception ex)
```

File: docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs [hunk 11]
```diff
@@ -2,43 +2,41 @@
...
+{
+    Console.WriteLine($"The decryption failed. {ex}");
+    throw;
```
Comment (IEvangelist) [inline docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs:41]:
```suggestion
```

File: docs/standard/security/decrypting-data.md [hunk 12]
```diff
@@ -21,18 +21,18 @@ Decryption is the reverse operation of encryption. For secret-key encryption, yo
...
 ```vb
 Dim aes As Aes = Aes.Create()
 Dim cryptStream As New CryptoStream(
-    myStream, aes.CreateDecryptor(key, iv), CryptoStreamMode.Read)
+    fileStream, aes.CreateDecryptor(key, iv), CryptoStreamMode.Read)
 ```
 
 ```csharp
 Aes aes = Aes.Create();
 CryptoStream cryptStream = new CryptoStream(
```
Comment (Youssef1313) [inline docs/standard/security/decrypting-data.md:34]: Is the lack of `using` intended here?

**Review by tdykstra — COMMENTED** [review]
LGTM but I added a request for @bartonjs to review.

File: docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs [hunk 13]
```diff
@@ -2,43 +2,40 @@
...
+    // it with the Aes class using the key and IV.
+    using CryptoStream cryptoStream = new(
+       fileStream,
+       aes.CreateDecryptor(key, iv),
+       CryptoStreamMode.Read);
+
+    // Read the stream.
+    using StreamReader decryptReader = new(cryptoStream);
+
+    var decryptedMessage = await decryptReader.ReadToEndAsync();
```
Comment (bartonjs) [inline docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs:34]: This doesn't meet the dotnet/runtime rules for when `var` is acceptable.

```suggestion
    string decryptedMessage = await decryptReader.ReadToEndAsync();
```

File: docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs [hunk 14]
```diff
@@ -2,43 +2,40 @@
...
 {
-    public static void Main(string[] args)
+    using FileStream fileStream = new("TestData.txt", FileMode.Open);
+
+    // Create a new instance of the default Aes implementation class
+    using Aes aes = Aes.Create();
+
+    // Reads IV value from beginning of the file.
+    byte[] iv = new byte[aes.IV.Length];
+    fileStream.Read(iv, 0, iv.Length);
```
Comment (bartonjs) [inline docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs:14]: While I believe it always works for FileStream, Stream does not guarantee that the array gets filled in one call.  Not sure if it's worth complicating the sample for it, but on the other hand, it does lead to people writing bad calls to Stream.Read where they make bad assumptions.

**Comment by bartonjs** [comment]
The title says that it's fixing a bug in the sample, but there are so many aesthetic changes that I can't see what may have actually needed to be changed.

I strongly recommend using braced usings for cryptographic/security code (and, honestly, for all code, but less strongly).  But they already weren't, so ::shrug::.

I'm not sure that the target-typed new enhances readability here.  (Personally, I only use it for long lines or multiple-argument generics)

The calls to CreateEncryptor and CreateDecryptor produce IDisposable values that never get disposed (CryptoStream doesn't dispose them).  If we're trying for perfection with our IDisposable-usings these'll get pulled out.

And, finally, the snippets lifted into the concept documents no longer match what the samples have.  If we want to hand wave and say that they're only highlighting the functional aspects of the code (which is why they're not showing the using statements), that's fine, but the encryption document shows a call to CreateEncryptor(key, iv) when the sample has changed to CreateEncryptor().

No major blockers, but these are my aesthetic observations.  It does look like the decrypt and encrypt samples are matched pairs.

File: docs/standard/security/decrypting-data.md (hunk 12 above)
Comment (IEvangelist) [reply docs/standard/security/decrypting-data.md:34]: In this context it is fine, as there rest of the code is omitted - we leave it to the reader's imagination. It is possible there there is a Try / Finally.

File: docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs (hunk 14 above)
Comment (IEvangelist) [reply docs/standard/security/snippets/decrypting-data/csharp/aes-decrypt.cs:14]: Good callout, I'm updating this now.

**Comment by IEvangelist** [comment]
Thanks for the review, @bartonjs and @tdykstra.

Yes, that was actually the bug that was being addressed here. I am following your advice now on the explicit scoping of the disposable implementations, part of the issue was not properly scoping these - careless of me really.

**Review by tdykstra — APPROVED** [review]
This looks better 😄

**Comment by aynur-safin** [comment]
You have done a great job ). Thank you.

I was late for the party, but maybe it will come in handy.
Encryption uses
`encryptWriter.WriteLine("Hello World!");`
but when decrypting
`await decryptReader.ReadToEndAsync();`
the resulting line contains a line break
`"Hello World!\r\n"`

When decrypting, a single asynchronous call is used
`await decryptReader.ReadToEndAsync();`
Can be used synchronous to keep everything in the same style, or alternatively make other calls asynchronous
`await encryptWriter.WriteLineAsync("Hello World!");`
etc.

**Pull request merged**